sdk/storage/azure-storage-internal-avro/src/main/java/com/azure/storage/internal/avro/implementation/schema/AvroSchema.java

```diff
@@ -0,0 +1,153 @@
+import static com.azure.storage.internal.avro.implementation.AvroConstants.Types.FIXED;
+import static com.azure.storage.internal.avro.implementation.AvroConstants.Types.FLOAT;
+import static com.azure.storage.internal.avro.implementation.AvroConstants.Types.INT;
+import static com.azure.storage.internal.avro.implementation.AvroConstants.Types.LONG;
+import static com.azure.storage.internal.avro.implementation.AvroConstants.Types.MAP;
+import static com.azure.storage.internal.avro.implementation.AvroConstants.Types.NULL;
+import static com.azure.storage.internal.avro.implementation.AvroConstants.Types.RECORD;
+import static com.azure.storage.internal.avro.implementation.AvroConstants.Types.STRING;
+import static com.azure.storage.internal.avro.implementation.AvroConstants.Types.UNION;
+
+/**
+ * An abstract class that represents an Avro schema that can return a generic result.
+ * AvroSchemas get placed on the AvroParserState stack and the AvroParser makes progress parsing the schemas on the
+ * stack as it is able to.
+ *
+ * @param <T> The result type of the schema.
+ * @see AvroParser#parse(ByteBuffer)
+ */
+public abstract class AvroSchema<T> {
+
+    protected final AvroParserState state;
+    private final Consumer<T> onResult;
+    protected boolean done = false;
+    protected T result;
+
+    /**
+     * Constructs a new Schema.
+     *
+     * @param state The state of the parser.
+     * @param onResult The result handler.
+     */
+    public AvroSchema(AvroParserState state, Consumer<T> onResult) {
+        this.state = state;
+        this.onResult = onResult;
+    }
+
+    /**
+     * Adds a schema to the state's stack.
+     * For complex types, this includes adding the initial child schema to make progress.
+     */
+    public abstract void add();
+
+    /**
+     * Makes some progress in parsing the type.
+     * For complex types, this is simply a no-op since primitive types do the job for them.
+     */
+    public abstract void progress();
+
+    /**
+     * @return Whether or not the schema is done. Also indicates that the result is ready.
+     */
+    public boolean isDone() {
+        return this.done;
+    }
+
+    /**
+     * @return Whether or not progress can be made for this schema.
+     * For complex types, this is always true since primitive types do the job for them.
+     */
+    public abstract boolean canProgress();
+
+    /**
+     * Calls the result handler.
+     */
+    public void publish() {
+        this.onResult.accept(result);
+    }
+
+    /**
+     * Gets the schema associated with the type.
+     *
+     * @param type The {@link AvroType type} that defines the schema.
+     * @param state {@link AvroParserState}
+     * @param onResult {@link Consumer}
+     * @return {@link AvroSchema}
+     *
+     * @see AvroType
+     */
+    public static AvroSchema getSchema(AvroType type, AvroParserState state, Consumer onResult) {
```

Comment:
Can I just add @SuppressWarnings("unchecked") here?
The build is in general super mad at me for casting unchecked

Reply:
Talked to Rick about it. I'm going to make AvroSchema ungeneric since it doesnt really fit the semantics and make the Consumers all Consumers of Objects and cast accordingly as I need Objects in Consumers.